### PR TITLE
CI: Use `tsd` for Testing Types

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

Uses the `tsd` package to run tests on types. This allows more nuanced type checking than `tsc`, and provides clear output when types don't match. ~~Not all of the `assertType`'s have been replaced, I've only done a few to demo how it works.~~

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating

### CODE REVIEW

File: typings/index.test-d.ts [hunk 1]
```diff
@@ -0,0 +1,1056 @@
...
+assertType<null>(loggedOutClient.readyAt);
+assertType<null>(loggedOutClient.readyTimestamp);
+assertType<string | null>(loggedOutClient.token);
+assertType<null>(loggedOutClient.uptime);
+assertType<null>(loggedOutClient.user);
+
+// Test type transformation:
+declare const assertType: <T>(value: T) => asserts value is T;
+declare const serialize: <T>(value: T) => Serialized<T>;
+declare const notPropertyOf: <T, P extends PropertyKey>(value: T, property: P & Exclude<P, keyof T>) => void;
```
Comment (vladfrangu) [inline typings/index.test-d.ts:707]: These should be moved at the top, since they're used above their definition too

File: .eslintignore [hunk 2]
```diff
@@ -0,0 +1 @@
+/typings/**/*
```
Comment (GodderE2D) [inline .eslintignore:1]:
```suggestion
/typings/**/*

```

File: package.json [hunk 3]
```diff
@@ -4,7 +4,7 @@
   "description": "A powerful library for interacting with the Discord API",
   "scripts": {
     "test": "npm run lint && npm run docs:test && npm run lint:typings && npm run test:typescript",
-    "test:typescript": "tsc --noEmit",
+    "test:typescript": "npx tsd",
```
Comment (iCrawl) [inline package.json:7]:
```suggestion
    "test:typescript": "tsc --noEmit && tsd",
```

At this point we might as well just do this and don't change the workflow at all.

**Comment by kyranet** [comment]
CI is failing.

Also, isn't the new file very similar to our `index.ts` file? Can't we re-use it somehow?

**Comment by suneettipirneni** [comment]
The index.ts file is the same file, it's just showing as a new file in the diff.

**Review by iCrawl — APPROVED** [review]

**Comment by iCrawl** [comment]
I'll just move forward to merge this, since it is a better alternative to maintaining some util functions ourselves and have proper testing this way.

**Pull request merged**